src/V3Increments.cpp

```diff
@@ -0,0 +1,222 @@
+    };
+
+    // STATE
+    int m_modIncrementsNum;  // Var name counter
+    InsertMode m_insMode;  // How to insert
+    AstNode* m_insStmtp;  // Where to insert statement
+    bool m_unsupportedHere;  // Used to detect where it's not supported yet
+
+    // METHODS
+    void insertBefore(AstNode* placep, AstNode* newp) {
+        AstNRelinker linker;
+        placep->unlinkFrBack(&linker);
+        newp->addNext(placep);
+        linker.relink(newp);
+    }
+
+private:
+    void insertBeforeStmt(AstNode* nodep, AstNode* newp) {
+        // Return node that must be visited, if any
+        // See also AstNode::addBeforeStmt; this predates that function
+        if (debug() >= 9) newp->dumpTree(cout, "-newstmt:");
+        UASSERT_OBJ(m_insStmtp, nodep, "Function not underneath a statement");
+        if (m_insMode == IM_BEFORE) {
+            // Add the whole thing before insertAt
+            if (debug() >= 9) newp->dumpTree(cout, "-newfunc:");
+            m_insStmtp->addHereThisAsNext(newp);
+        } else if (m_insMode == IM_AFTER) {
+            m_insStmtp->addNextHere(newp);
+        } else if (m_insMode == IM_WHILE_PRECOND) {
+            AstWhile* whilep = VN_CAST(m_insStmtp, While);
+            UASSERT_OBJ(whilep, nodep, "Insert should be under WHILE");
+            whilep->addPrecondsp(newp);
+        } else {
+            nodep->v3fatalSrc("Unknown InsertMode");
+        }
+        m_insMode = IM_AFTER;
+        m_insStmtp = newp;
+    }
+
+    virtual void visit(AstWhile* nodep) VL_OVERRIDE {
+        // Special, as statements need to be put in different places
+        // Preconditions insert first just before themselves (the normal
+        // rule for other statement types)
+        m_insStmtp = NULL;  // First thing should be new statement
+        iterateAndNextNull(nodep->precondsp());
+        // Conditions insert first at end of precondsp.
+        m_insMode = IM_WHILE_PRECOND;
+        m_insStmtp = nodep;
+        iterateAndNextNull(nodep->condp());
+        // Body insert just before themselves
+        m_insStmtp = NULL;  // First thing should be new statement
+        iterateAndNextNull(nodep->bodysp());
+        iterateAndNextNull(nodep->incsp());
+        // Done the loop
+        m_insStmtp = NULL;  // Next thing should be new statement
+    }
+
+    virtual void visit(AstNodeFor* nodep) VL_OVERRIDE {
+        nodep->v3fatalSrc(
+            "For statements should have been converted to while statements in V3Begin.cpp");
+    }
+
+    virtual void visit(AstNodeStmt* nodep) VL_OVERRIDE {
+        if (!nodep->isStatement()) {
+            iterateChildren(nodep);
+            return;
+        }
+        m_insMode = IM_BEFORE;
+        m_insStmtp = nodep;
+        iterateChildren(nodep);
+        m_insStmtp = NULL;  // Next thing should be new statement
+    }
+
+    virtual void visit(AstLogAnd* nodep) VL_OVERRIDE {
+        // PreAdd and Post add not yet supported after &&
```

Comment:
I think there's a lot of logical operators which will break. Perhaps better to check for what IS legal - which would be whatever you test in your regression test.

Reply:
Wouldn't that be actually harder? What I mean that by checking for `AstLogAnd` (and possibly others) I can just assume that everything that is down the tree should be rejected and just reject the whole branch, if I am looking for nodes to be accepted wouldn't I have to somehow accept whole branches of the syntax tree? For example to be sure that whatever I accept is not inside `AstLogAnd `(or other that we wish not to accept).

I will of course be happy to implement it like this if you have some pointers about "how deep" into the tree should I look for the nodes to be accepted (to be sure that everything in them and everything that wraps them is supported).